crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <openssl/opensslconf.h>
+#ifdef FIPS_MODE
+NON_EMPTY_TRANSLATION_UNIT
+#else
+# include <stdio.h>
+# include <string.h>
+# include <openssl/evp.h>
+# include <openssl/rand_drbg.h>
+# include "internal/cryptlib.h"
+# include "crypto/bn.h"
+# include "internal/thread_once.h"
+# include "../rand/rand_local.h"
+
+typedef struct dsa_drbg_ctx_st
+{
+    const unsigned char *x;
+    const unsigned char *h;
+    size_t rlen;
+} DSA_DRBG_CTX;
+
+static void *dsa_drbg_ctx_new(OPENSSL_CTX *ctx)
+{
+    DSA_DRBG_CTX *dsa_drbg_ctx = OPENSSL_zalloc(sizeof(*dsa_drbg_ctx));
+    return dsa_drbg_ctx;
+}
+
+static void dsa_drbg_ctx_free(void *dsa_drbg_ctx)
+{
+    OPENSSL_free(dsa_drbg_ctx);
+}
+
+static const OPENSSL_CTX_METHOD dsa_drbg_ctx_method = {
+    dsa_drbg_ctx_new,
+    dsa_drbg_ctx_free
+};
+
+static size_t get_entropy(RAND_DRBG *drbg, unsigned char **pout,
+                          int entropy, size_t min_len, size_t max_len,
+                          int prediction_resistance)
+{
+    DSA_DRBG_CTX *dsa_drbg_ctx = openssl_ctx_get_data(drbg->libctx, OPENSSL_CTX_DSA_DRBG_INDEX,
+                                                      &dsa_drbg_ctx_method);
```

Comment:
Creating a dedicated 'library' context just for passing some context data to the `get_entropy` and `get_nonce` callbacks seems a little bit overengineered IMHO. Also, it doesn't solve your problem of retrieving the context data, because the `drbg` structure is opaque outside of `crypto/rand`. (And this is **not** going to change.)

Currently, it is not possible to pass context data to the callbacks.

        int RAND_DRBG_set_callbacks(RAND_DRBG *drbg,
                                    RAND_DRBG_get_entropy_fn get_entropy,
                                    RAND_DRBG_cleanup_entropy_fn cleanup_entropy,
                                    RAND_DRBG_get_nonce_fn get_nonce,
                                    RAND_DRBG_cleanup_nonce_fn cleanup_nonce)
        
An option would be to add this as a new feature by adding two accessors

        int RAND_DRBG_set_callback_data(RAND_DRBG *drbg, void * data
                                    RAND_DRBG_cleanup_data_fn cleanup_data)
        
        void *RAND_DRBG_get_callback_data(RAND_DRBG *drbg);

(Note the extra `cleanup_data` callback for freeing the allocated data.) The documentation will emphasize that access to the data is not synchronized, it is primarily intended to be used only for private DRBGs which are not shared by different threads. (Alternatively, the DRBG's owner needs to provide the synchronization.)

Reply:
What I suggested, already exists: please take a look at `RAND_DRBG_{set,get}_ex_data()`:

https://github.com/openssl/openssl/blob/5cf5342c116d643bc810ded630d1848a40439bd7/include/openssl/rand_drbg.h#L127-L128

Reply:
For examples how it can be used, take a look at the callbacks for the known answer tests:

https://github.com/openssl/openssl/blob/5cf5342c116d643bc810ded630d1848a40439bd7/test/drbg_cavs_test.c#L34-L53

Reply:
The code used to be like this before [5cf5342](https://github.com/openssl/openssl/pull/9223/commits/5cf5342c116d643bc810ded630d1848a40439bd7)
Please view https://github.com/openssl/openssl/pull/9223#discussion_r300306010

mattcaswell suggested to use OPENSSL_CTX to store the data, so I modified the code according to his suggestions.

Reply:
Ok, thanks for the hint. Many things were changed recently by the heavy replumbing and I am not yet familiar with all the details and pitfalls of the architectural innovations. Nevertheless, the current solution looks heavyweight and overengineered to me.  I will have to take another look and try to better understand the details.